python/uds.py

```diff
@@ -0,0 +1,717 @@
+    self.service_id = service_id
+    self.error_code = error_code
+
+class InvalidServiceIdError(Exception):
+  pass
+
+class InvalidSubFunctioneError(Exception):
+  pass
+
+_negative_response_codes = {
+    0x00: 'positive response',
+    0x10: 'general reject',
+    0x11: 'service not supported',
+    0x12: 'sub-function not supported',
+    0x13: 'incorrect message length or invalid format',
+    0x14: 'response too long',
+    0x21: 'busy repeat request',
+    0x22: 'conditions not correct',
+    0x24: 'request sequence error',
+    0x25: 'no response from subnet component',
+    0x26: 'failure prevents execution of requested action',
+    0x31: 'request out of range',
+    0x33: 'security access denied',
+    0x35: 'invalid key',
+    0x36: 'exceed numebr of attempts',
+    0x37: 'required time delay not expired',
+    0x70: 'upload download not accepted',
+    0x71: 'transfer data suspended',
+    0x72: 'general programming failure',
+    0x73: 'wrong block sequence counter',
+    0x78: 'request correctly received - response pending',
+    0x7e: 'sub-function not supported in active session',
+    0x7f: 'service not supported in active session',
+    0x81: 'rpm too high',
+    0x82: 'rpm too low',
+    0x83: 'engine is running',
+    0x84: 'engine is not running',
+    0x85: 'engine run time too low',
+    0x86: 'temperature too high',
+    0x87: 'temperature too low',
+    0x88: 'vehicle speed too high',
+    0x89: 'vehicle speed too low',
+    0x8a: 'throttle/pedal too high',
+    0x8b: 'throttle/pedal too low',
+    0x8c: 'transmission not in neutral',
+    0x8d: 'transmission not in gear',
+    0x8f: 'brake switch(es) not closed',
+    0x90: 'shifter lever not in park',
+    0x91: 'torque converter clutch locked',
+    0x92: 'voltage too high',
+    0x93: 'voltage too low',
+}
+
+class UdsClient():
+  def __init__(self, panda, tx_addr, rx_addr=None, bus=0, timeout=10, debug=False):
+    self.panda = panda
+    self.bus = bus
+    self.tx_addr = tx_addr
+    if rx_addr == None:
+      if tx_addr < 0xFFF8:
+        self.rx_addr = tx_addr+8
+      elif tx_addr > 0x10000000 and tx_addr < 0xFFFFFFFF:
+        self.rx_addr = (tx_addr & 0xFFFF0000) + (tx_addr<<8 & 0xFF00) + (tx_addr>>8 & 0xFF)
+      else:
+        raise ValueError("invalid tx_addr: {}".format(tx_addr))
+
+    self.tx_queue = Queue()
+    self.rx_queue = Queue()
+    self.timeout = timeout
+    self.debug = debug
+
+    self.can_reader_t = threading.Thread(target=self._isotp_thread, args=(self.debug,))
+    self.can_reader_t.daemon = True
+    self.can_reader_t.start()
+
+  def _isotp_thread(self, debug):
+    try:
+      rx_frame = {"size": 0, "data": "", "idx": 0, "done": True}
+      tx_frame = {"size": 0, "data": "", "idx": 0, "done": True}
+
+      # allow all output
+      self.panda.set_safety_mode(0x1337)
+      # clear tx buffer
+      self.panda.can_clear(self.bus)
+      # clear rx buffer
+      self.panda.can_clear(0xFFFF)
```

Comment:
Is that what this does?

Reply:
https://github.com/commaai/panda/blob/4972376de36aff020694d06d2458d73a5f4da829/board/main.c#L499